noir/noir-repo/aztec_macros/src/transforms/note_interface.rs

```diff
@@ -0,0 +1,510 @@
+            note_struct.fields.push((
+                ident("header"),
+                make_type(UnresolvedTypeData::Named(
+                    chained_dep!("aztec", "note", "note_header", "NoteHeader"),
+                    vec![],
+                    false,
+                )),
+            ));
+        }
+        for (field_ident, field_type) in note_struct.fields.iter() {
+            note_fields.push((
+                field_ident.0.contents.to_string(),
+                field_type.typ.to_string().replace("plain::", ""),
+            ));
+        }
+
+        let note_interface_impl_span = trait_impl.object_type.span;
+
+        // Generate the serialize_content method as
+        //
+        // fn serialize_content(self: {}) -> [Field; NOTE_SERIALIZED_LEN] {
+        //   [self.note_field1 as Field, self.note_field2.to_field()...]
+        // }
+        //
+        // It assumes every struct field can be converted either via the to_field() trait (structs) or cast as Field (integers)
+
+        if !check_trait_method_implemented(trait_impl, "serialize_content") {
+            let note_serialize_content_fn = generate_note_serialize_content(
+                &note_type,
+                &note_fields,
+                &note_serialized_len,
+                note_interface_impl_span,
+            )?;
+            trait_impl.items.push(TraitImplItem::Function(note_serialize_content_fn));
+        }
+
+        // Generate the deserialize_content method as
+        //
+        // fn deserialize_content(serialized_note: [Field; NOTE_SERILIZED_LEN]) -> Self {
+        //     NoteType {
+        //        note_field1: serialized_note[0] as Field,
+        //        note_field2: NoteFieldType2::from_field(serialized_note[1])...
+        //     }
+        // }
+        // It assumes every note field is stored in an individual serialized field,
+        // and can be converted to the original type via the from_field() trait (structs) or cast as Field (integers)
+
+        if !check_trait_method_implemented(trait_impl, "deserialize_content") {
+            let note_deserialize_content_fn = generate_note_deserialize_content(
+                &note_type,
+                &note_fields,
+                &note_serialized_len,
+                note_interface_impl_span,
+            )?;
+            trait_impl.items.push(TraitImplItem::Function(note_deserialize_content_fn));
+        }
+
+        // Automatically generate a struct that represents the note's serialization metadata, and an auxiliary function to retrieve it
+        // The struct looks like this
+        //
+        // NoteTypeFields {
+        //     field1: FieldSelector { index: 0, offset: 0, length: 32 },
+        //     field2: FieldSelector { index: 1, offset: 0, length: 32 },
+        //     ...
+        // }
+        //
+        // It assumes each field occupies an entire field and its serialized in definition order
```

Comment:
I think we should only automatically generate this method if we have also autogenerated the `serialize` method, otherwise the user could've defined serialization in a different order, and it'd silently screw up getters.

Reply:
Good point, done!